### PR TITLE
Switch over function input matching to use Params

### PR DESCRIPTION
**Pull request by CodaFi**

### CODE REVIEW

**Comment by CodaFi** [comment]
@swift-ci please smoke test

**Comment by CodaFi** [comment]
@swift-ci please smoke test

File: lib/Sema/CSSimplify.cpp [hunk 1]
```diff
@@ -982,6 +982,118 @@ static bool matchFunctionRepresentations(FunctionTypeRepresentation rep1,
...
+  SmallVector<Identifier, 2> argLabelsScratch;
+  bool hasTrailingClosure = false;
+  std::tie(callee, calleeLevel, argLabels, hasTrailingClosure) =
+    getCalleeDeclAndArgs(*this, locator, argLabelsScratch);
+  // FIXME: If we're unable to dig out a callee, defer to match types.  This
+  // occurs e.g. when we bind overloads.  These code paths occur too early to
+  // set resolvedOverloads, so require special handling to bind any latent
+  // type variables.
+  if (!callee) {
+    // Peephole type variable matching.
```
Comment (DougGregor) [inline lib/Sema/CSSimplify.cpp]: This is basically cloning the first part of `matchTypes()`. Why do we need the peephole?

File: lib/Sema/CSSimplify.cpp [hunk 2]
```diff
@@ -982,6 +982,118 @@ static bool matchFunctionRepresentations(FunctionTypeRepresentation rep1,
...
+  
+  SmallVector<bool, 4> defaultMap;
+  computeDefaultMap(argType, callee, calleeLevel, defaultMap);
+  
+  // Match up the call arguments to the parameters.
+  MatchCallArgumentListener listener;
+  SmallVector<ParamBinding, 4> parameterBindings;
+  if (constraints::matchCallArguments(type2, type1, defaultMap,
+                                      hasTrailingClosure,
+                                      false,
```
Comment (DougGregor) [inline lib/Sema/CSSimplify.cpp]: `/*allowFixes=*/false`, please

**Review by DougGregor — APPROVED** [review]
Nifty. This is a nice step forward

**Comment by CodaFi** [comment]
@swift-ci please smoke test

**Comment by CodaFi** [comment]
@swift-ci please smoke test

**Comment by CodaFi** [comment]
@swift-ci please test

**Comment by CodaFi** [comment]
@swift-ci please test

**Comment by CodaFi** [comment]
⛵️

**Pull request merged**